torch_glow/src/PyTorchModelLoader.cpp

```diff
@@ -25,47 +28,354 @@ glow::Type ptTypeToGlowType(const at::CompleteTensorType &ptType) {
+
+  glow::NodeValue input = getGlowNodeValue(inputs[0]);
+  glow::ReluNode *glowNode = f_->createRELU("relu", input);
+  addGlowNodeValue(outputs[0], glowNode->getResult());
+}
+
+void PyTorchModelLoader::loadConvolution(const torch::jit::Node *ptNode) {
+  auto inputs = ptNode->inputs();
+  auto outputs = ptNode->outputs();
+  assert(inputs.size() == 12);
+  assert(outputs.size() == 1);
+
+  // Glow expects conv inputs to be in NHWC but PyTorch keeps them in NCHW so we
+  // tranpose them.
+  glow::NodeValue input = getGlowNodeValue(inputs[0]);
+  input = f_->createTranspose("conv_input_transposed", input, NCHW2NHWC);
+  glow::ShapeNHWC inputShape(input.dims());
+
+  // Glow expects conv weights to be in CRSK but PyTorch keeps them in CKRS
+  // so we tranpose them. C - output_depth, R - filter_height, S -
+  // filter_width, K - input_depth.
+  glow::NodeValue weights = getGlowNodeValue(inputs[1]);
+  weights = f_->createTranspose("conv_weights_transposed", weights, NCHW2NHWC);
+  glow::ShapeNHWC weightsShape(weights.dims());
+
+  // If a bias was provided then use it otherwise create a 0 bias.
+  glow::NodeValue bias;
+  if (hasGlowNodeValue(inputs[2])) {
+    bias = getGlowNodeValue(inputs[2]);
+  } else {
+    glow::Tensor biasT(glow::ElemKind::FloatTy, {weightsShape.n});
+    biasT.zero();
+    glow::Constant *biasConstant =
+        f_->getParent()->createConstant("conv_bias", std::move(biasT));
+    bias = biasConstant->getOutput();
   }
-  case at::aten::div: {
-    assert(inputs.size() == 2);
-    glow::NodeValue rhs = valueMap_.at(inputs[0]);
-    glow::NodeValue lhs = valueMap_.at(inputs[1]);
-    glowNode = f_->createDiv("div", rhs, lhs);
-    valueMap_[outputs[0]] = glowNode->getNthResult(glow::DivNode::ResultIdx);
-    break;
+
+  glow::Handle<int32_t> stridesHandle =
+      getGlowConstantHandle<int32_t>(inputs[3]);
+  std::vector<uint32_t> strides =
+      expandParamIfNeeded<int32_t, uint32_t>(stridesHandle, 2);
+
+  glow::Handle<int32_t> padsHandle = getGlowConstantHandle<int32_t>(inputs[4]);
+  uint32_t pad = static_cast<uint32_t>(contractParamIfNeeded(padsHandle));
+  std::vector<uint32_t> pads = {pad, pad, pad, pad};
+
+  glow::Handle<int32_t> dilationHandle =
+      getGlowConstantHandle<int32_t>(inputs[5]);
+  uint32_t dilation =
+      static_cast<uint32_t>(contractParamIfNeeded(dilationHandle));
+
+  // Don't support transposed convolutions yet.
+  glow::Handle<bool> transposedHandle = getGlowConstantHandle<bool>(inputs[6]);
+  assert(transposedHandle.size() == 1 && !transposedHandle.raw(0) &&
+         "Transposed convolutions not supported");
+  (void)transposedHandle;
+
+  glow::Handle<int32_t> groupsHandle =
+      getGlowConstantHandle<int32_t>(inputs[8]);
+  assert(groupsHandle.size() == 1);
+  uint32_t groups = groupsHandle.raw(0);
+
+  std::vector<uint32_t> kernels = {static_cast<uint32_t>(weightsShape.h),
+                                   static_cast<uint32_t>(weightsShape.w)};
+
```

Comment:
btw where to see docs about other 3 inputs? (since we only deal with 9 here)

Reply:
I referenced the code https://fburl.com/6m7fgaf9

Reply:
oh okay, then the other 3 does not seem to be related.